pkg/ring/lifecycler.go

```diff
@@ -242,12 +252,59 @@ func (i *Lifecycler) getTokens() []uint32 {
 	return i.tokens
 }
 
+func (i *Lifecycler) flushTokensToFile() {
+	if i.cfg.TokenFileDir == "" {
+		return
+	}
+	tokenFilePath := path.Join(i.cfg.TokenFileDir, tokensFileName)
+	f, err := os.OpenFile(tokenFilePath, os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "error in creating token file", "err", err)
+		return
+	}
+
+	b := make([]byte, 4*len(i.tokens))
+	for idx, token := range i.tokens {
+		binary.BigEndian.PutUint32(b[idx*4:], token)
+	}
+
+	if _, err = f.Write(b); err != nil {
+		level.Error(util.Logger).Log("msg", "error in writing token file", "err", err)
+		return
+	}
+}
+
+func (i *Lifecycler) getTokensFromFile() ([]uint32, error) {
+	tokenFilePath := path.Join(i.cfg.TokenFileDir, tokensFileName)
+	b, err := ioutil.ReadFile(tokenFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	if len(b) == 0 {
+		return nil, nil
+	} else if len(b)%4 != 0 {
+		return nil, errors.New("token data is not 4 byte aligned")
+	}
+
+	numTokens := len(b) >> 2
+	tokens := make([]uint32, 0, numTokens)
+	for i := 0; i < numTokens; i++ {
+		tokens = append(tokens, binary.BigEndian.Uint32(b[i<<2:]))
+	}
+
+	return tokens, nil
+}
+
 func (i *Lifecycler) setTokens(tokens []uint32) {
 	tokensOwned.WithLabelValues(i.RingName).Set(float64(len(tokens)))
 
 	i.stateMtx.Lock()
 	defer i.stateMtx.Unlock()
 	i.tokens = tokens
+	i.flushTokensToFile()
```

Comment:
Instead of loking while flushing tokens to disk, you may pass `tokens` to `flushTokensToFile()`, like this:

```
i.stateMtx.Lock()
i.tokens = tokens
i.stateMtx.Unlock()

i.flushTokensToFile(tokens)
```

Reply:
Yup makes sense. I had thought with the lock we can assure that tokens were into the disk. But I guess we don't need that guarantee.